dimagi/commcare-hq
# Update location through user api

## Product Description

This PR add logic to update location for mobile workers through user api. 

- get mobile worker through api will also return `primary_location` and `locations`
- create new mobile worker and update existing mobile worker now accept two new fields: `primary_location` and `locations`.
- if the two fields is not included in the json object, nothing will happen.
- if the two fields is included, `primary_location` is set to empty string, and `locations` is set to empty array, then all locations and primary location will be removed from that user.

Criteria on updating locations:
- the two new fields must be provided together
- `primary_location` must be one of `locations`
- all location must be valid

Updated wiki: (Please don't click on Update)
List Mobile Workers API: https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/edit-v2/2143957296
User Creation: https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/edit-v2/2143957350
User Edit: https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/edit-v2/2143957433

## Technical Summary

Ticket: https://dimagi.atlassian.net/browse/SAAS-15722
This PR introduces changes to the CommCare User API, mimicking the UI of change location for a mobile worker: it will always take two fields, `primary_location` (a string that represents location id of the primary location) and `locations` (an array of location ids) together.
The update logic ensures that both fields must be provided together, with validation ensuring that the primary location is included in the list of locations, all location are valid. 
Then I will take the two field to one location_obj dict, and pass it to `update` function, which will be called in both create mobile worker and edit mobile worker functions.


## Safety Assurance

### Safety story

I tested the changes locally and on staging. 

Validation has been added to prevent incorrect or incomplete updates.

I think I maintained backward compatibility, if this two field is not included, then the api will still work just as how it works before without editing locations of mobile workers. But if user is using the returned data of GET request for further process, I did return two more fields: `locations` and `primary_location`, will this be a concern?

### Automated test coverage


corehq/apps/api/tests/test_user_resources.py
corehq/apps/api/tests/test_user_updates.py

### QA Plan


QA Ticket: https://dimagi.atlassian.net/browse/QA-7085

QA Plan:
- Test List Mobile Workers api, and verify, `primary_location` and `locations` are now included in the returned results
- Test Create Mobile Worker api and verify mobile workers with locations can be created
- Test Edit Mobile Worker api and verify mobile workers's location can be updated
      - When no location field included, then locations won't be updated
      - When only `primary_location` or `locations` included, then error is raised
      - When invalid location passed, then error is raised
      - When `primary_location` is not in the list of `locations`, error is raised
- Prevent Location Updates in user_data:
Ensure that location information cannot be overwritten by putting `commcare_location_ids` and `commcare_location_id` in user_data. The update will be rejected and both the location and user_data remains unchanged. 
-  A regression test on other user apis

### Rollback instructions




### Labels & Review
